backend/controllers/chatController.js

```diff
@@ -0,0 +1,154 @@
+        }
+
+        // validate timestamp format
+        if (!is_valid_timestamp(timestamp)) {
+            throw new Error("Timestamp requested is invalid");
+        } 
+        // validate timestamp logic
+        if (new Date(timestamp) > Date.now()) {
+            throw new Error("Timestamp requested is in the future");
+        }
+
+        const chat = await Chat.findById({ _id: chat_id });
+        const messages = await load_message_buffer(chat.message, buffer_size, timestamp);
+        const polls = await load_poll_buffer(chat.polls, buffer_size, timestamp);
+        return sort_by_timestamp(messages, polls, buffer_size);
+    } catch (err) {
+        throw new Error(`${err.message}`);
+    }
+}
+
+async function load_message_buffer(message_ids, buffer_size, timestamp) {
+    try {
+        let messages = [];
+        if (message_ids.length === 0) return messages;
+        // return x messages based on timestamps provided
+        messages = await Message.find({
+            _id: { $in: message_ids },
+            createdAt: { $lte: new Date(timestamp) }, // Filter messages before the given timestamp
+        })
+        .limit(buffer_size)
+        .populate("author", "user_name has_account")
+        .select("author content createdAt")
+        return messages;
+    } catch (err) {
+        throw new Error(`Failed to load message buffer: ${err.message}`);
+    }
+}
+
+async function load_poll_buffer(poll_ids, buffer_size, timestamp) {
+    try {
+        let polls = [];
+        if (poll_ids.length === 0) return polls;
+        // return x polls based on timestamps provided
+        polls = await Poll.find({
+            _id: { $in: poll_ids},
+            createdAt: { $lte: new Date(timestamp) }, // Filter polls before the given timestamp
+        })
+        .limit(buffer_size)
+        .populate("options", "title vote_count")
+        .populate("users_voted", "user_name")
+        .select("title options users_voted createdAt") // no need to get chat._id
+        return polls;
+    } catch (err) {
+        throw new Error(`Failed to load poll buffer: ${err.message}`);
+    }
+}
+
+function sort_by_timestamp(messages, polls, buffer_size) {
+    try {
+        const chat_history = [...messages, ...polls];
+        if (chat_history.length == 0) return chat_history;
+        // Sort by createdAt in ascending order (latest -> oldest)
+        chat_history.sort((a, b) => new Date(a.createdAt) - new Date(b.createdAt)); 
+        if (buffer_size >= chat_history.length) return chat_history;
+        return chat_history.slice(buffer_size);
+    } catch (err) {
+        throw new Error(`Failed to sort chat_history: ${err.message}`);
+    }
+}
+
+function is_valid_timestamp(timestamp) {
+    if (timestamp === null) {
+        return false;
+    }
+    if (typeof timestamp === "Date") {
+        return !isNaN(timestamp.getTime());
+    }
+    const date = new Date(timestamp);
+    return !isNaN(date.getTime()) && !isNaN(date.getMinutes());
+}
```

Comment:
Move this under load chat since it's the first function called. Or move it under the verification function if you make it.

Reply:
> Will make another commit to refactor all these functions following this structure